### PR TITLE
feat(backend): support ES indices for user and workspace

### PR DESCRIPTION
**Pull request by asf-algoo**
See #4147.

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)
- [x] Make sure that:
  - if there are modifications in the Tracim configuration files (eg. `development.ini`), they are documented in `backend/doc/setting.md`
  - any migration process required for existing instances is documented
  - relevant people for these changes are notified

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

**Testing method**

- launch third-party services with `docker-compose up -d` from backend directory
- edit `development.ini` to set:

```ini
search.engine = elasticsearch
search.elasticsearch.index_alias_prefix = tracim
```

- activate python virtual env
- check that `tracimcli search index-create` creates 3 indices prefixed by `tracim-(content|user|workspace)`. You can check by browsing to [http://localhost:9200/_stats](http://localhost:9200/_stats)
- check that `tracimcli search index-populate` finishes correctly
- - check that `tracimcli search index-upgrade-experimental` finishes correctly
- check that `tracimcli search index-drop` deletes the 3 indices prefixed by `tracim-(content|user|workspace)`.

### CODE REVIEW

File: backend/tracim_backend/command/search.py [hunk 1]
```diff
@@ -28,10 +28,10 @@ def _index_all_contents(self) -> None:
         nb_index_errors = index_contents_result.get_nb_index_errors()
         nb_contents_to_index = index_contents_result.get_nb_contents_to_index()
         if nb_index_errors == 0:
-            print("All {} content where indexed".format(nb_contents_to_index))
+            print("All {} content(s) were indexed".format(nb_contents_to_index))
```
Comment (raphj) [inline backend/tracim_backend/command/search.py:31]: I would expect All to always be followed by a plural, and all 1 content look a bit weird.

```suggestion
            print("{} content(s) were indexed".format(nb_contents_to_index))
```

File: backend/tracim_backend/command/search.py [hunk 2]
```diff
@@ -28,10 +28,10 @@ def _index_all_contents(self) -> None:
...
         else:
             print(
-                "Warning ! {}/{} contents cannot be indexed properly.".format(
+                "Warning ! {}/{} content(s) cannot be indexed properly.".format(
```
Comment (raphj) [inline backend/tracim_backend/command/search.py:34]:
```suggestion
                "Warning! Failed to index {}/{} content(s)".format(
```

File: backend/tracim_backend/command/search.py [hunk 3]
```diff
@@ -134,4 +134,4 @@ def take_app_action(self, parsed_args: argparse.Namespace, app_context: AppEnvir
         )
         print("delete index")
         self.search_api.delete_index()
-        print("Index where deleted")
+        print("Index were deleted")
```
Comment (raphj) [inline backend/tracim_backend/command/search.py:137]:
```suggestion
        print("Indices were deleted")
```

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 4]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
-        """
-        # FIXME BS 2019-06-10: Load ES model only when ES search (see #1892)
-        from tracim_backend.lib.search.elasticsearch_search.es_models import IndexedContent
-
-        # INFO - G.M - 2019-05-15 - alias migration mechanism to allow easily updatable index.
+        # INFO - G.M - 2019-05-15 - alias migration mechanism to allow easily updateable index.
         # from https://github.com/elastic/elasticsearch-dsl-py/blob/master/examples/alias_migration.py
         # Configure index with our indexing preferences
-        logger.info(self, "Create index settings ...")
+        logger.info(self, "Create ES indices...")
```
Comment (raphj) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:70]:
```suggestion
        logger.info(self, "Creating ES indices...")
```

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 5]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
-        next_index = self.index_document_pattern_template.replace(
-            "{index_alias}", self.index_document_alias
-        ).replace("{date}", datetime.now().strftime("%Y%m%d%H%M%S%f"))
+        for parameters in self._get_indices_parameters():
+            new_index_name = self._get_index_name(parameters)
 
-        logger.info(self, 'create new index "{}"'.format(next_index))
-        # create new index, it will use the settings from the template
-        self.es.indices.create(index=next_index)
+            logger.info(self, 'create new index "{}"'.format(new_index_name))
```
Comment (raphj) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:147]:
```suggestion
            logger.info(self, 'Creating new index "{}"'.format(new_index_name))
```

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 6]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
-                        {
-                            "remove": {
-                                "alias": self.index_document_alias,
-                                "index": self.index_document_pattern,
-                            }
-                        },
-                        {"add": {"alias": self.index_document_alias, "index": next_index}},
-                    ]
-                }
+                'set alias "{}" to point on index "{}"'.format(parameters.alias, new_index_name),
```
Comment (raphj) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:164]:
```suggestion
                'Setting alias "{}" to point on index "{}"'.format(parameters.alias, new_index_name),
```

**Review by raphj — APPROVED** [review]
lgtm, but I'm not quite familiar with this part of Tracim anyway

File: backend/tracim_backend/command/search.py [hunk 7]
```diff
@@ -134,4 +134,4 @@ def take_app_action(self, parsed_args: argparse.Namespace, app_context: AppEnvir
         )
         print("delete index")
         self.search_api.delete_index()
```
Comment (inkhey) [inline backend/tracim_backend/command/search.py:136]:
```suggestion
        self.search_api.delete_index()
```
Plural form may be better now, we do delete multiples index, no ?

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 8]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
                 )
             ]
         )
-        self.index_document_pattern = config.SEARCH__ELASTICSEARCH__INDEX_PATTERN_TEMPLATE.format(
-            date="*", index_alias=config.SEARCH__ELASTICSEARCH__INDEX_ALIAS
-        )
-        self.index_document_pattern_template = config.SEARCH__ELASTICSEARCH__INDEX_PATTERN_TEMPLATE
-        self.index_document_alias = config.SEARCH__ELASTICSEARCH__INDEX_ALIAS
 
     def create_index(self) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:66]:
```suggestion
    def create_indices(self) -> None:
```
?

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 9]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
+                            "alias": parameters.alias,
+                            "index": self._get_index_pattern(parameters),
+                        }
+                    },
+                    {"add": {"alias": parameters.alias, "index": index_name}},
+                ]
+            }
+        )
 
     def refresh_index(self) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:113]:
```suggestion
    def refresh_indices(self) -> None:
```
?

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 10]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
+        refresh index to obtain up to date information instead of relying on
+        periodical refresh, useful for automated tests.
         see https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-refresh.html
         """
-
-        self.es.indices.refresh(self.index_document_alias)
+        for parameters in self._get_indices_parameters():
+            self.es.indices.refresh(parameters.alias)
 
     def delete_index(self) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:122]:
```suggestion
    def delete_indice(self) -> None:
```

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py [hunk 11]
```diff
@@ -45,128 +62,114 @@ def __init__(self, session: Session, current_user: typing.Optional[User], config
...
-        self.es.indices.delete(self.index_document_pattern, allow_no_indices=True)
-        self.es.indices.delete_template(self.index_document_alias)
+        for parameters in self._get_indices_parameters():
+            index_pattern = self._get_index_pattern(parameters)
+            logger.info(self, "deleting indices whose name matches {}".format(index_pattern))
+            self.es.indices.delete(index_pattern, allow_no_indices=True)
+            self.es.indices.delete_template(parameters.alias)
 
-    def migrate_index(self, move_data=True, update_alias=True) -> None:
+    def migrate_index(self) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py:132]:
```suggestion
    def migrate_indices(self) -> None:
```
?

File: backend/tracim_backend/lib/search/search.py [hunk 12]
```diff
@@ -49,7 +49,7 @@ def create_index(self) -> None:
         pass
 
     @abstractmethod
-    def migrate_index(self, move_data=True, update_alias=True) -> None:
+    def migrate_index(self) -> None:
```
Comment (inkhey) [inline backend/tracim_backend/lib/search/search.py:52]:
```suggestion
    def migrate_indices(self) -> None:
```
?

**Review by inkhey — APPROVED** [review]
seems ok to me, just need to be consistent with or without plural form about index/indices .

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 8 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py]: Done.

File: backend/tracim_backend/lib/search/search.py (hunk 12 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/search.py]: Done.

File: backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py (hunk 9 above)
Comment (grignards) [reply backend/tracim_backend/lib/search/elasticsearch_search/elasticsearch_search.py]: Done.

**Pull request merged**